test/gtest/uct/test_pending.cc

```diff
@@ -93,96 +132,97 @@ class test_uct_pending : public uct_test {
-            --req->countdown;
-            return UCS_INPROGRESS;
+        /* We expect that message sent from pending callback will arrive
+         * before the one sent from the completion callback. */
+        if (hdr == PENDING_HDR) {
+            pend_received = true;
+            EXPECT_FALSE(*comp_received);
+        } else if (hdr == COMP_HDR) {
+            *comp_received = true;
+            EXPECT_TRUE(pend_received);
+        } else {
+            EXPECT_EQ(AM_HDR, hdr);
         }
 
-        status = uct_ep_am_short(req->ep, 0, test_pending_hdr, &req->data,
-                                 sizeof(req->data));
-        if (status == UCS_OK) {
-            pending_delete(req);
-        }
-        return status;
+        return UCS_OK;
     }
 
-    static ucs_status_t pending_send_op_simple(uct_pending_req_t *self) {
+    static void completion_cb(uct_completion_t *self, ucs_status_t c_status) {
+        am_completion_t *comp = ucs_container_of(self, am_completion_t, uct);
+
+        EXPECT_UCS_OK(c_status);
+
+        ucs_status_t status = uct_ep_am_short(comp->ep, AM_ID, COMP_HDR,
+                                              NULL, 0);
+        EXPECT_TRUE(!UCS_STATUS_IS_ERR(status) ||
+                    (status == UCS_ERR_NO_RESOURCE));
+    }
 
-        pending_send_request_t *req = ucs_container_of(self, pending_send_request_t, uct);
+    static ucs_status_t pending_send_op(uct_pending_req_t *self) {
+
+        pending_send_request_t *req = ucs_container_of(self,
+                                                       pending_send_request_t,
+                                                       uct);
         ucs_status_t status;
 
-        status = uct_ep_am_short(req->ep, 0, test_pending_hdr, &req->data,
+        status = uct_ep_am_short(req->ep, AM_ID, PENDING_HDR, &req->data,
                                  sizeof(req->data));
         if (status == UCS_OK) {
-            req->countdown ++;
+            req->countdown++;
             n_pending--;
             req->active = 0;
-            //ucs_warn("dispatched %p idx %d total %d", req->ep, req->id, req->countdown);
         }
         return status;
     }
 
-    static ucs_status_t pending_send_op_bcopy(uct_pending_req_t *self) {
+    static ucs_status_t pending_send_op_count(uct_pending_req_t *self) {
 
-        pending_send_request_t *req = ucs_container_of(self, pending_send_request_t, uct);
-        ssize_t packed_len;
+        pending_send_request_t *req = ucs_container_of(self,
+                                                       pending_send_request_t,
+                                                       uct);
+        ucs_status_t status;
 
-        packed_len = uct_ep_am_bcopy(req->ep, 0, mapped_buffer::pack, req->buf, 0);
-        if (packed_len > 0) {
-            req->countdown ++;
-            n_pending--;
-            req->active = 0;
-            return UCS_OK;
+        if (req->countdown > 0) {
+            --req->countdown;
```

Comment:
it's weird that countdown goes down here, but goes up in pending_send_op
can we unite pending_send_op and pending_send_op_count?